File: .github/scripts/validate_bob_review_result.py
#!/usr/bin/env python3
"""
Validate Bob review artifacts for the V3 fully agentic flow.

This script does not analyze code.
It only verifies that Bob produced the required decision output contract.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path


REQUIRED_TOP_LEVEL_KEYS = {
    "status",
    "confidence",
    "summary",
    "findings",
    "positives",
    "block_merge",
}


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate Bob review result")
    parser.add_argument("--json-result", required=True, help="Path to Bob JSON result")
    parser.add_argument("--markdown-result", required=True, help="Path to Bob Markdown review")
    args = parser.parse_args()

    json_path = Path(args.json_result)
    markdown_path = Path(args.markdown_result)

    if not json_path.exists():
        print("Missing Bob JSON result")
        return 2

    if not markdown_path.exists():
        print("Missing Bob Markdown review")
        return 2

    data = json.loads(json_path.read_text(encoding="utf-8"))
    missing = sorted(REQUIRED_TOP_LEVEL_KEYS - set(data.keys()))
    if missing:
        print(f"Bob JSON result is missing required keys: {', '.join(missing)}")
        return 2

    if not isinstance(data["findings"], list):
        print("Bob JSON result field 'findings' must be a list")
        return 2

    if not isinstance(data["positives"], list):
        print("Bob JSON result field 'positives' must be a list")
        return 2

    if not isinstance(data["block_merge"], bool):
        print("Bob JSON result field 'block_merge' must be boolean")
        return 2

    status = str(data["status"]).upper()
    if status not in {"PASS", "FAIL", "NEEDS_ATTENTION"}:
        print("Bob JSON result field 'status' must be PASS, FAIL, or NEEDS_ATTENTION")
        return 2

    print("Bob review artifacts are structurally valid")

    if data["block_merge"]:
        print("Bob decided to block merge")
        return 1

    print("Bob approved or did not block merge")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

# Made with Bob

File: .github/llama-agent/validate-bob-review-result.js
import fs from 'node:fs';
import path from 'node:path';

const REQUIRED_TOP_LEVEL_KEYS = [
  'status',
  'confidence',
  'summary',
  'findings',
  'positives',
  'block_merge'
];

function parseArgs(argv) {
  const args = {
    jsonResult: '.github/artifacts/bob-review.json',
    markdownResult: '.github/artifacts/bob-review.md'
  };

  for (let i = 2; i < argv.length; i += 1) {
    const key = argv[i];
    const value = argv[i + 1];
    if (!value) {
      continue;
    }

    if (key === '--json-result') {
      args.jsonResult = value;
      i += 1;
    } else if (key === '--markdown-result') {
      args.markdownResult = value;
      i += 1;
    }
  }

  return args;
}

function fail(message, code) {
  process.stdout.write(`${message}\n`);
  process.exit(code);
}

function main() {
  const args = parseArgs(process.argv);
  const jsonPath = path.resolve(args.jsonResult);
  const markdownPath = path.resolve(args.markdownResult);

  if (!fs.existsSync(jsonPath)) {
    fail('Missing Bob JSON result', 2);
  }

  if (!fs.existsSync(markdownPath)) {
    fail('Missing Bob Markdown review', 2);
  }

  const data = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
  const missing = REQUIRED_TOP_LEVEL_KEYS.filter((key) => !(key in data));

  if (missing.length > 0) {
    fail(`Bob JSON result is missing required keys: ${missing.join(', ')}`, 2);
  }

  if (!Array.isArray(data.findings)) {
    fail(`Bob JSON result field 'findings' must be a list`, 2);
  }

  if (!Array.isArray(data.positives)) {
    fail(`Bob JSON result field 'positives' must be a list`, 2);
  }

  if (typeof data.block_merge !== 'boolean') {
    fail(`Bob JSON result field 'block_merge' must be boolean`, 2);
  }

  const status = String(data.status).toUpperCase();
  if (!['PASS', 'FAIL', 'NEEDS_ATTENTION'].includes(status)) {
    fail(`Bob JSON result field 'status' must be PASS, FAIL, or NEEDS_ATTENTION`, 2);
  }

  process.stdout.write('Bob review artifacts are structurally valid\n');

  if (data.block_merge) {
    process.stdout.write('Bob decided to block merge\n');
    process.exit(1);
  }

  process.stdout.write('Bob approved or did not block merge\n');
}

main();

// Made with Bob
